### PR TITLE
Added incrementally increasing quarantine

### PR DESCRIPTION
**Pull request by michaeljfazio**
Linearly increases the quarantine period relative to the total number of lifetime strikes a node has recorded against it. This change helps to prevent dead-peers being gossiped on the network by ignoring them for longer and longer durations until they are effectively no-longer gossip candidates.

### CODE REVIEW

**Comment by michaeljfazio** [comment]
Should probably add logic to cap/reset the quarantine period to prevent legitimate nodes that have experienced significant downtime from being forgotten for too long.

File: src/policy.rs [hunk 1]
```diff
@@ -58,11 +59,13 @@ impl Policy for DefaultPolicy {
         // if the node is already quarantined
         if let Some(since) = node.logs().quarantined() {
             let duration = since.elapsed().unwrap();
+            let quarantine_duration =
+                DEFAULT_QUARANTINE_DURATION.mul(node.record().lifetime_strikes);
```
Comment (NicolasDP) [inline src/policy.rs:63]: we need to have some form of boundary though. something like `max(quarantine_duration, MAX_QUARANTINE_DURATION)`. so we don't keep quarantined object for ever.

**Review by NicolasDP — CHANGES_REQUESTED** [review]

**Comment by NicolasDP** [comment]
thanks for the update here. So indeed we need to have a better management of the quarantine policy in order to be more conservative. A solution could be to add a quarantine history/records within the `DefaultPolicy` object: something like:

```rust
#[derive(Default)]
pub struct DefaultPolicy {
    history: LRU<NodeId, NodeHistory>,
}

struct NodeHistory {
    lifetime_strikes: usize,
    last_update: std::time::Instant,
}
```

The idea is to keep a boundary on the number of quarantined nodes so we don't keep things forever. Also because it is a LRU, at some point it will be forgotten. Also if the NodeHistory has not been updated since _certain amount of time_ then we can clean the slate too or reset the lifetime_strikes counter.

**Comment by michaeljfazio** [comment]
> thanks for the update here. So indeed we need to have a better management of the quarantine policy in order to be more conservative. A solution could be to add a quarantine history/records within the `DefaultPolicy` object: something like:
> 
> ```rust
> #[derive(Default)]
> pub struct DefaultPolicy {
>     history: LRU<NodeId, NodeHistory>,
> }
> 
> struct NodeHistory {
>     lifetime_strikes: usize,
>     last_update: std::time::Instant,
> }
> ```
> 
> The idea is to keep a boundary on the number of quarantined nodes so we don't keep things forever. Also because it is a LRU, at some point it will be forgotten. Also if the NodeHistory has not been updated since _certain amount of time_ then we can clean the slate too or reset the lifetime_strikes counter.

I like it. Is using the LRU cache going to be too non-deterministic with regards to purging history data though? If for instance the cache were set to 10,000 and there were 1000 reachable nodes and 9000 non-reachable nodes on the network, then your individual history lists would still grow unbounded and never be cleared (at least not within a practical time frame). So, in theory you would end up with memory consumption issues here. I think.... ?

Maybe a bounded FIFO queue for history would be better? Certainly more deterministic and easier to reason about at least.

**Comment by NicolasDP** [comment]
> Is using the LRU cache going to be too non-deterministic with regards to purging history data

this is why we also add the `last_update` and check it with `Instant::now()`. we can then purge data as we see fit if certain time has passed but was not garbage collected. This will prevent the case the cache is bigger than the number of nodes.
also that way we get GC for free, if it is updated, it is kept longer, if not updated (the node is no longer used/active) it is then likely to be GCed at some point and so to have a clean slate.

**Comment by michaeljfazio** [comment]
> > Is using the LRU cache going to be too non-deterministic with regards to purging history data
> 
> this is why we also add the `last_update` and check it with `Instant::now()`. we can then purge data as we see fit if certain time has passed but was not garbage collected. This will prevent the case the cache is bigger than the number of nodes.
> also that way we get GC for free, if it is updated, it is kept longer, if not updated (the node is no longer used/active) it is then likely to be GCed at some point and so to have a clean slate.

I was attempting to reason about where the memory was being restricted by the LRU. Now that I've through it through a bit better I can see how the nested LRU make sense as it is in-effect a bounded FIFO queue as I suggested in the previous comment. I totally agree with you now. I'll prepare the change accordingly.

**Comment by NicolasDP** [comment]
Are you okay we close this PR?

**Comment by michaeljfazio** [comment]
> Are you okay we close this PR?

Go for it @NicolasDP 👍

**Pull request closed without merging**